hazelcast/src/main/java/com/hazelcast/crdt/pncounter/PNCounter.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.crdt.pncounter;
+
+import com.hazelcast.config.PNCounterConfig;
+import com.hazelcast.core.DistributedObject;
+import com.hazelcast.crdt.MutationDisallowedException;
+import com.hazelcast.internal.cluster.ClusterService;
+
+/**
+ * PN (Positive-Negative) CRDT counter.
+ * <p>
+ * The counter supports adding and subtracting values as well as
+ * retrieving the current counter value.
+ * Each replica of this counter can perform operations locally without
+ * coordination with the other replicas, thus increasing availability.
+ * The counter guarantees that whenever two nodes have received the
+ * same set of updates, possibly in a different order, their state is
+ * identical, and any conflicting updates are merged automatically.
+ * If no new updates are made to the shared state, all nodes that can
+ * communicate will eventually have the same data.
+ * <p>
+ * Each replica is identified by a UUID. This identifier should be
+ * unique and there should not be any other replica with the same ID, even
+ * if the member with that ID is no longer alive.
+ * <p>
+ * The updates to this counter are applied locally when invoked on a
+ * CRDT replica. A replica can be any hazelcast instance which is not a
+ * client or a lite member. The number of replicas in the cluster is
+ * determined by the {@link PNCounterConfig#getReplicaCount()} configuration
+ * value.
+ * <p>
+ * When invoking updates from non-replica instance, the invocation is remote.
+ * This may lead to indeterminate state - the update may be applied but the
+ * response has not been received. In this case, the caller will be notified
+ * with a {@link com.hazelcast.spi.exception.TargetDisconnectedException}.
```

Comment:
I think the invocation system can also throw `MemberLeftException` if the invocation is between hz instances.

Reply:
True, will update.

Reply:
Added comment